File: mvps/dotnet/Services/VoteService.cs
using System.Collections.Concurrent;
using BeerVote.Dtos;
using BeerVote.Errors;
using BeerVote.Models;

namespace BeerVote.Services;

public sealed class VoteService
{
    private readonly UserService _userService;
    private readonly BeerCatalog _beerCatalog;

    private readonly ConcurrentDictionary<Guid, Vote> _votesById = new();
    private readonly ConcurrentDictionary<Guid, Guid> _voteIdByUserId = new();

    public VoteService(UserService userService, BeerCatalog beerCatalog)
    {
        _userService = userService;
        _beerCatalog = beerCatalog;
    }

    public VoteResponse Cast(CreateVoteRequest? request)
    {
        if (request?.UserId is null)
        {
            throw new ApiException(400, "user_id is required");
        }

        var userId = request.UserId.Value;

        if (!_userService.Exists(userId))
        {
            throw new ApiException(404, "user not found");
        }

        var beer = _beerCatalog.FindById(request.BeerId)
            ?? throw new ApiException(400, "beer_id must be one of the nominees (1..10)");

        var voteId = Guid.NewGuid();
        var vote = new Vote(voteId, userId, beer.Id, DateTimeOffset.UtcNow);

        if (!_voteIdByUserId.TryAdd(userId, voteId))
        {
            throw new ApiException(409, "user has already voted");
        }

        _votesById[voteId] = vote;
        return new VoteResponse(vote.Id, vote.UserId, vote.BeerId, beer.Name, vote.VotedAt);
    }

    public ICollection<Vote> AllVotes() => _votesById.Values;
}

File: mvps/java-26/src/main/java/com/meetup/beervote/vote/VoteService.java
package com.meetup.beervote.vote;

import com.meetup.beervote.beer.Beer;
import com.meetup.beervote.beer.BeerCatalog;
import com.meetup.beervote.dto.ApiDtos.CreateVoteRequest;
import com.meetup.beervote.dto.ApiDtos.VoteResponse;
import com.meetup.beervote.error.ApiException;
import com.meetup.beervote.user.UserService;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.Collection;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class VoteService {

    private final UserService userService;
    private final BeerCatalog beerCatalog;

    /** All votes keyed by vote id. */
    private final ConcurrentHashMap<UUID, Vote> votesById = new ConcurrentHashMap<>();

    /** Ensures one vote per user (userId → voteId). */
    private final ConcurrentHashMap<UUID, UUID> voteIdByUserId = new ConcurrentHashMap<>();

    public VoteService(UserService userService, BeerCatalog beerCatalog) {
        this.userService = userService;
        this.beerCatalog = beerCatalog;
    }

    public VoteResponse cast(CreateVoteRequest request) {
        if (request == null || request.userId() == null) {
            throw new ApiException(400, "user_id is required");
        }

        if (!userService.exists(request.userId())) {
            throw new ApiException(404, "user not found");
        }

        Beer beer = beerCatalog.findById(request.beerId())
                .orElseThrow(() -> new ApiException(400, "beer_id must be one of the nominees (1..10)"));

        UUID voteId = UUID.randomUUID();
        Vote vote = new Vote(voteId, request.userId(), beer.getId(), Instant.now());

        UUID existing = voteIdByUserId.putIfAbsent(request.userId(), voteId);
        if (existing != null) {
            throw new ApiException(409, "user has already voted");
        }

        votesById.put(voteId, vote);
        return new VoteResponse(vote.id(), vote.userId(), vote.beerId(), beer.getName(), vote.votedAt());
    }

    public Collection<Vote> allVotes() {
        return votesById.values();
    }
}
